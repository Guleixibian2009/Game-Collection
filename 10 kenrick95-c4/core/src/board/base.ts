import { Player } from '../player'
import { getMockPlayerAction } from '../utils'

export enum BoardPiece {
  EMPTY,
  PLAYER_1,
  PLAYER_2,
  DRAW,
}
export class BoardBase {
  static readonly ROWS: number = 6
  static readonly COLUMNS: number = 7
  static readonly PLAYER_1_COLOR: string = '#ef453b'
  static readonly PLAYER_2_COLOR: string = '#0059ff'
  static readonly PIECE_STROKE_STYLE: string = 'black'
  static readonly MASK_COLOR: string = '#d8d8d8'
  static CANVAS_HEIGHT: number
  static CANVAS_WIDTH: number
  static PIECE_RADIUS: number
  static MASK_X_BEGIN: number
  static MASK_Y_BEGIN: number
  static MESSAGE_WIDTH: number
  static MESSAGE_X_BEGIN: number
  static MESSAGE_Y_BEGIN: number
  static SCALE: number

  map: Array<Array<number>>
  protected winnerBoardPiece: BoardPiece

  constructor() {
    this.map = []
    this.winnerBoardPiece = BoardPiece.EMPTY
    this.initConstants()
    this.reset()
  }

  reset() {
    this.map = []
    for (let i = 0; i < BoardBase.ROWS; i++) {
      this.map.push([])
      for (let j = 0; j < BoardBase.COLUMNS; j++) {
        this.map[i].push(BoardPiece.EMPTY)
      }
    }
    this.winnerBoardPiece = BoardPiece.EMPTY
  }

  initConstants() {
    BoardBase.CANVAS_HEIGHT = BoardBase.SCALE * 480
    BoardBase.CANVAS_WIDTH = BoardBase.SCALE * 640
    BoardBase.PIECE_RADIUS = BoardBase.SCALE * 25
    BoardBase.MASK_X_BEGIN =
      Math.max(
        0,
        BoardBase.CANVAS_WIDTH -
          (3 * BoardBase.COLUMNS + 1) * BoardBase.PIECE_RADIUS
      ) / 2
    BoardBase.MASK_Y_BEGIN =
      Math.max(
        0,
        BoardBase.CANVAS_HEIGHT -
          (3 * BoardBase.ROWS + 1) * BoardBase.PIECE_RADIUS
      ) / 2
    BoardBase.MESSAGE_WIDTH = BoardBase.SCALE * 400
    BoardBase.MESSAGE_X_BEGIN =
      (BoardBase.CANVAS_WIDTH - BoardBase.MESSAGE_WIDTH) / 2
    BoardBase.MESSAGE_Y_BEGIN = BoardBase.SCALE * 20
  }

  /**
   * @returns is the action succesfully applied
   * @param player current player
   * @param column the colum in which the player want to drop a piece
   */
  async applyPlayerAction(player: Player, column: number): Promise<boolean> {
    const { success: actionSuccessful, map: nextState } = getMockPlayerAction(
      this.map,
      player.boardPiece,
      column
    )

    this.map = nextState
    // this.debug()

    return actionSuccessful
  }

  debug() {
    console.log(this.map.map((row) => row.join(' ')).join('\n'))
  }

  getWinner(): BoardPiece {
    if (this.winnerBoardPiece !== BoardPiece.EMPTY) {
      return this.winnerBoardPiece
    }
    const direction = [
      [0, -1],
      [0, 1],
      [-1, -1],
      [-1, 0],
      [-1, 1],
      [1, -1],
      [1, 0],
      [1, 1],
    ]
    const isWinningSequence = (
      i: number,
      j: number,
      playerPiece: BoardPiece,
      dir: Array<number>,
      count: number
    ): boolean => {
      if (count >= 4) {
        return true
      }
      if (
        i < 0 ||
        j < 0 ||
        i >= BoardBase.ROWS ||
        j >= BoardBase.COLUMNS ||
        this.map[i][j] !== playerPiece
      ) {
        return false
      }
      return isWinningSequence(
        i + dir[0],
        j + dir[1],
        playerPiece,
        dir,
        count + 1
      )
    }
    let countEmpty = 0
    for (let i = 0; i < BoardBase.ROWS; i++) {
      for (let j = 0; j < BoardBase.COLUMNS; j++) {
        const playerPiece = this.map[i][j]
        if (playerPiece !== BoardPiece.EMPTY) {
          for (let k = 0; k < direction.length; k++) {
            const isWon = isWinningSequence(
              i + direction[k][0],
              j + direction[k][1],
              playerPiece,
              direction[k],
              1
            )
            if (isWon) {
              return (this.winnerBoardPiece = playerPiece)
            }
          }
        } else {
          countEmpty++
        }
      }
    }
    if (countEmpty === 0) {
      return (this.winnerBoardPiece = BoardPiece.DRAW)
    }

    return BoardPiece.EMPTY
  }

  protected getPlayerColor(boardPiece: BoardPiece): string {
    switch (boardPiece) {
      case BoardPiece.PLAYER_1:
        return BoardBase.PLAYER_1_COLOR
      case BoardPiece.PLAYER_2:
        return BoardBase.PLAYER_2_COLOR
      default:
        return 'transparent'
    }
  }
}
