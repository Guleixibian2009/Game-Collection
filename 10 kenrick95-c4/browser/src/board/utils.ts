import { Board } from './index'

export function drawCircle(
  context: CanvasRenderingContext2D,
  { x = 0, y = 0, r = 0, fillStyle = '', strokeStyle = '' }
) {
  context.save()
  context.fillStyle = fillStyle
  context.strokeStyle = strokeStyle
  context.beginPath()
  context.arc(x, y, r, 0, 2 * Math.PI, false)
  context.fill()
  context.restore()
}
/**
 * @see http://stackoverflow.com/a/11770000/917957
 * @param context Canvas 2D Context
 * @param board   current board
 */
export function drawMask(board: Board) {
  const context = board.context
  context.save()
  context.fillStyle = Board.MASK_COLOR
  context.beginPath()
  const doubleRadius = 2 * Board.PIECE_RADIUS
  const tripleRadius = 3 * Board.PIECE_RADIUS
  for (let y = 0; y < Board.ROWS; y++) {
    for (let x = 0; x < Board.COLUMNS; x++) {
      context.arc(
        tripleRadius * x + Board.MASK_X_BEGIN + doubleRadius,
        tripleRadius * y + Board.MASK_Y_BEGIN + doubleRadius,
        Board.PIECE_RADIUS,
        0,
        2 * Math.PI
      )
      context.rect(
        tripleRadius * x + Board.MASK_X_BEGIN + 2 * doubleRadius,
        tripleRadius * y + Board.MASK_Y_BEGIN,
        -2 * doubleRadius,
        2 * doubleRadius
      )
    }
  }
  context.fill()
  context.restore()
}

export function clearCanvas(board: Board) {
  board.context.clearRect(0, 0, board.canvas.width, board.canvas.height)
}
