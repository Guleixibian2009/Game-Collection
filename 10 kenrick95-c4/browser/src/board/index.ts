import { BoardBase, BoardPiece } from '@kenrick95/c4/src/board'
import { Player } from '@kenrick95/c4/src/player'
import { onresize, animationFrame } from '@kenrick95/c4/src/utils'
import { drawMask, drawCircle, clearCanvas } from './utils'

export class Board extends BoardBase {
  canvas: HTMLCanvasElement
  context: CanvasRenderingContext2D

  constructor(canvas: HTMLCanvasElement) {
    super()
    this.canvas = canvas
    this.context = <CanvasRenderingContext2D>canvas.getContext('2d')
    this.getBoardScale()
    this.initConstants()
    this.reset()
    this.onresize()
  }

  getBoardScale() {
    return window.innerWidth < 640
      ? (BoardBase.SCALE = 0.5)
      : (BoardBase.SCALE = 1.0)
  }

  onresize() {
    let prevBoardScale = BoardBase.SCALE
    onresize().add(() => {
      this.getBoardScale()
      if (prevBoardScale !== BoardBase.SCALE) {
        prevBoardScale = BoardBase.SCALE
        this.initConstants()
        clearCanvas(this)
        this.render()
      }
    })
  }

  reset() {
    super.reset()
    if (this.canvas) {
      clearCanvas(this)
      this.render()
    }
  }

  initConstants() {
    super.initConstants()
    if (this.canvas) {
      /**
       * Scale the canvas to make it look sharper on hi-dpi devices
       * https://www.html5rocks.com/en/tutorials/canvas/hidpi/
       */
      const dpr = self.devicePixelRatio || 1
      this.canvas.width = Board.CANVAS_WIDTH * dpr
      this.canvas.height = Board.CANVAS_HEIGHT * dpr
      this.context.scale(dpr, dpr)
      this.canvas.style.width = Board.CANVAS_WIDTH + 'px'
      this.canvas.style.height = Board.CANVAS_HEIGHT + 'px'
    }
  }

  private async animateAction(
    newRow: number,
    column: number,
    boardPiece: BoardPiece
  ): Promise<void> {
    const fillStyle = this.getPlayerColor(boardPiece)
    let currentY = 0
    const doAnimation = async () => {
      clearCanvas(this)
      drawCircle(this.context, {
        x:
          3 * BoardBase.PIECE_RADIUS * column +
          BoardBase.MASK_X_BEGIN +
          2 * BoardBase.PIECE_RADIUS,
        y: currentY + BoardBase.MASK_Y_BEGIN + 2 * BoardBase.PIECE_RADIUS,
        r: BoardBase.PIECE_RADIUS,
        fillStyle: fillStyle,
        strokeStyle: BoardBase.PIECE_STROKE_STYLE,
      })
      this.render()
      currentY += BoardBase.PIECE_RADIUS
    }
    while (newRow * 3 * BoardBase.PIECE_RADIUS >= currentY) {
      await animationFrame()
      doAnimation()
    }
  }

  render() {
    drawMask(this)
    for (let y = 0; y < BoardBase.ROWS; y++) {
      for (let x = 0; x < BoardBase.COLUMNS; x++) {
        drawCircle(this.context, {
          x:
            3 * BoardBase.PIECE_RADIUS * x +
            BoardBase.MASK_X_BEGIN +
            2 * BoardBase.PIECE_RADIUS,
          y:
            3 * BoardBase.PIECE_RADIUS * y +
            BoardBase.MASK_Y_BEGIN +
            2 * BoardBase.PIECE_RADIUS,
          r: BoardBase.PIECE_RADIUS,
          fillStyle: this.getPlayerColor(this.map[y][x]),
          strokeStyle: BoardBase.PIECE_STROKE_STYLE,
        })
      }
    }
  }

  async applyPlayerAction(player: Player, column: number): Promise<boolean> {
    if (
      this.map[0][column] !== BoardPiece.EMPTY ||
      column < 0 ||
      column >= BoardBase.COLUMNS
    ) {
      return false
    }

    let isColumnEverFilled = false
    let row = 0
    for (let i = 0; i < BoardBase.ROWS - 1; i++) {
      if (this.map[i + 1][column] !== BoardPiece.EMPTY) {
        isColumnEverFilled = true
        row = i
        break
      }
    }
    if (!isColumnEverFilled) {
      row = BoardBase.ROWS - 1
    }

    await this.animateAction(row, column, player.boardPiece)

    // reflect player's action to the map
    this.map[row][column] = player.boardPiece
    this.debug()

    await animationFrame()
    this.render()
    return true
  }
}
