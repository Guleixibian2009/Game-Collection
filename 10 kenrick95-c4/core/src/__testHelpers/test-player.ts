import { Player } from '../player'
import { BoardPiece } from '../board'

export class TestPlayer extends Player {
  getActionPromiseResolver: null | ((column: number) => void)

  /**
   * This promise acts to wait at `doAction` until there is a `getActionPromiseResolver`
   */
  getActionWaiting: null | Promise<void> = null
  getActionWaitingResolver: null | (() => void)

  constructor(boardPiece: BoardPiece) {
    super(boardPiece)
    this.getActionPromiseResolver = null

    this.getActionWaitingResolver = null
    this.renewPromise()
  }
  renewPromise() {
    this.getActionWaiting = new Promise(
      (resolve) => (this.getActionWaitingResolver = resolve)
    )
  }
  async getAction() {
    return new Promise<number>((resolve) => {
      this.getActionPromiseResolver = resolve

      if (this.getActionWaitingResolver) {
        this.getActionWaitingResolver()
      }
    })
  }
  async doAction(column: number) {
    await this.getActionWaiting

    if (this.getActionPromiseResolver) {
      this.getActionPromiseResolver(column)
    }
    this.renewPromise()
  }
}
