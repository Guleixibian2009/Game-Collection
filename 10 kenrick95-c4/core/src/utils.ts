import { BoardPiece, BoardBase } from './board/base'

export const BIG_POSITIVE_NUMBER: number = 10 ** 9 + 7
export const BIG_NEGATIVE_NUMBER: number = -BIG_POSITIVE_NUMBER

export function showMessage(message: string = ''): void {
  const messageDOM: Element | null = document.querySelector('.message')

  if (!messageDOM) return console.error('Message DOM is null!')

  messageDOM.classList.remove('hidden')

  const messageContentDOM: Element | null = document.querySelector(
    '.message-body-content'
  )

  if (!messageContentDOM)
    return console.error('Message body content DOM is null!')

  messageContentDOM.innerHTML = message

  const messageDismissDOM: Element | null = document.querySelector(
    '.message-body-dismiss'
  )

  if (!messageDismissDOM)
    return console.error('Message body dismiss DOM is null!')

  const dismissHandler = (): void => {
    messageDOM.classList.add('invisible')
    messageDOM.addEventListener('transitionend', (): void => {
      messageDOM.classList.add('hidden')
      messageDOM.classList.remove('invisible')
    })

    messageDismissDOM.removeEventListener('click', dismissHandler)
  }

  messageDismissDOM.addEventListener('click', dismissHandler)
}

/**
 *
 * @param {{ x: number; y: number }} coord The Coordinates of the point to be checked.
 * @param {number} columnXBegin The X-Coordinate of N-th column.
 * @param {number} radius The radius of a piece.
 */
export function isCoordOnColumn(
  coord: { x: number; y: number },
  columnXBegin: number,
  radius: number
): boolean {
  return (coord.x - columnXBegin) * (coord.x - columnXBegin) <= radius * radius
}

export function getColumnFromCoord(coord: { x: number; y: number }): number {
  for (let i: number = 0; i < BoardBase.COLUMNS; i++)
    if (
      isCoordOnColumn(
        coord,
        3 * BoardBase.PIECE_RADIUS * i +
          BoardBase.MASK_X_BEGIN +
          2 * BoardBase.PIECE_RADIUS,
        BoardBase.PIECE_RADIUS
      )
    )
      return i

  return -1
}

export function getRandomColumnNumber(): number {
  return Math.floor(Math.random() * BoardBase.COLUMNS)
}

export function choose(choice: Array<any>): any {
  return choice[Math.floor(Math.random() * choice.length)]
}

/**
 * @see https://esdiscuss.org/topic/promises-async-functions-and-requestanimationframe-together
 */
export function animationFrame(): Promise<Function> {
  let resolve: Function | null = null
  const promise: Promise<Function> = new Promise(
    (r: Function): Function => (resolve = r)
  )

  if (resolve) window.requestAnimationFrame(resolve)

  return promise
}

export function clone(array: Array<Array<any>>): Array<Array<any>> {
  const arr: Array<Array<any>> = []

  for (let i: number = 0; i < array.length; i++) arr[i] = array[i].slice()

  return arr
}

export function getMockPlayerAction(
  map: Array<Array<number>>,
  boardPiece: BoardPiece,
  column: number
): {
  success: boolean
  map: Array<Array<number>>
} {
  const clonedMap: Array<Array<any>> = clone(map)

  if (
    clonedMap[0][column] !== BoardPiece.EMPTY ||
    column < 0 ||
    column >= BoardBase.COLUMNS
  )
    return {
      success: false,
      map: clonedMap,
    }

  let isColumnEverFilled: boolean = false
  let row: number = 0
  for (let i: number = 0; i < BoardBase.ROWS - 1; i++)
    if (clonedMap[i + 1][column] !== BoardPiece.EMPTY) {
      isColumnEverFilled = true
      row = i

      break
    }

  if (!isColumnEverFilled) row = BoardBase.ROWS - 1

  clonedMap[row][column] = boardPiece

  return {
    success: true,
    map: clonedMap,
  }
}

/**
 * From Mozilla Developer Network
 * https://developer.mozilla.org/en-US/docs/Web/Events/resize
 */
export function onresize(): { add: Function } {
  const callbacks: Array<Function> = []
  let running: boolean = false

  // Fired on resize event.
  function resize() {
    if (!running) {
      running = true

      if (window.requestAnimationFrame)
        window.requestAnimationFrame(runCallbacks)
      else setTimeout(runCallbacks, 66)
    }
  }

  // Run the actual callbacks.
  function runCallbacks() {
    callbacks.forEach((callback: Function): void => {
      callback()
    })

    running = false
  }

  // Adds callback to loop.
  function addCallback(callback: Function): void {
    if (callback) callbacks.push(callback)
  }

  return {
    // Public method to add additional callback.
    add: (callback: Function) => {
      if (!callbacks.length) window.addEventListener('resize', resize)

      addCallback(callback)
    },
  }
}
