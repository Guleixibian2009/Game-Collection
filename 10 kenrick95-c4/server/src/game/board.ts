import { BoardBase } from '@kenrick95/c4/src/board'

export class ServerBoard extends BoardBase {}
