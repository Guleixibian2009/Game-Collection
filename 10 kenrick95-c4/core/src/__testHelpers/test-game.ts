import { GameBase } from '../game'
import { BoardBase } from '../board'
import { Player } from '../player'

export class TestGame extends GameBase {
  afterMoveResolve: null | (() => void) = null
  afterMovePromise: null | Promise<void> = null

  constructor(players: Array<Player>, board: BoardBase) {
    super(players, board)
    this.renewAfterMovePromise()
  }

  waitingForMove() {
    // no-op
  }
  beforeMoveApplied() {
    // no-op
  }
  afterMove() {
    if (this.afterMoveResolve) {
      this.afterMoveResolve()
    }
    this.renewAfterMovePromise()
  }

  renewAfterMovePromise() {
    this.afterMovePromise = new Promise(
      (resolve) => (this.afterMoveResolve = resolve)
    )
  }
}
