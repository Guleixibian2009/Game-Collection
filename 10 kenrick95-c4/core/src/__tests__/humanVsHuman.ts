import { TestGame } from '../__testHelpers/test-game'
import { BoardPiece, BoardBase } from '../board'
import { TestPlayer } from '../__testHelpers/test-player'
import { clone } from '../utils'

describe('PlayerHuman vs PlayerHuman', () => {
  const players = [
    new TestPlayer(BoardPiece.PLAYER_1),
    new TestPlayer(BoardPiece.PLAYER_2),
  ]
  test('Basic test', async () => {
    const board = new BoardBase()
    const game = new TestGame(players, board)
    game.start()
    expect(board.map).toMatchSnapshot()
    await players[0].doAction(3)
    await game.afterMovePromise
    expect(board.map).toMatchSnapshot()
    await players[1].doAction(3)
    await game.afterMovePromise
    expect(board.map).toMatchSnapshot()
  })
  test('Cannot move on a filled column', async () => {
    const board = new BoardBase()
    const game = new TestGame(players, board)
    game.start()
    await players[0].doAction(3)
    await players[1].doAction(3)
    await players[0].doAction(3)
    await players[1].doAction(3)
    await players[0].doAction(3)
    await players[1].doAction(3)
    await game.afterMovePromise
    const currentMap = clone(board.map)
    expect(game.currentPlayerId).toEqual(0)
    expect(board.map).toMatchSnapshot()
    await players[0].doAction(3)
    expect(board.map).toEqual(currentMap)
    expect(game.currentPlayerId).toEqual(0)
    expect(board.map).toMatchSnapshot()
  })
})
