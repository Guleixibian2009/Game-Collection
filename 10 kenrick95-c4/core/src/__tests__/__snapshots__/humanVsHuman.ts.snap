// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerHuman vs PlayerHuman Basic test 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
]
`;

exports[`PlayerHuman vs PlayerHuman Basic test 2`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
]
`;

exports[`PlayerHuman vs PlayerHuman Basic test 3`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
]
`;

exports[`PlayerHuman vs PlayerHuman Cannot move on a filled column 1`] = `
Array [
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
]
`;

exports[`PlayerHuman vs PlayerHuman Cannot move on a filled column 2`] = `
Array [
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    2,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
  ],
]
`;
